### ISSUE TITLE
Deadlock when reusing redis connection

### ISSUE THREAD

**Issue by stephankaag** [dv/redis-semaphore#18]
Works:

```
  i = 0
  threads = [1, 2].collect do
    Thread.new do
      Redis::Semaphore.new("foo", :redis => Redis.new(:db => 15)).lock do
        i += 1
      end
    end
  end
  threads.each &:join
  i.should == 2
```

Deadlocks:

```
  i = 0
  threads = [1, 2].collect do
    Thread.new do
      Redis::Semaphore.new("foo", :redis => @redis).lock do
        i += 1
      end
    end
  end
  threads.each &:join
  i.should == 2
```

**Comment by stephankaag** [comment]
Looks good to me. @dv ?

**Comment by dv** [comment]
Thanks for reminding me @stephankaag, I'll find some free time this week and work through the open issues on this gem!

**Comment by kimrgrey** [comment]
Yeah, using of the same redis connection in, for example, different sidekiq workers is not good idea. It will leat to endless lock of this threads. Think, it's better to clarify this fact in README.

**Comment by dv** [comment]
I agree, due to the blocking nature of some usecases of `redis-semaphore`, reusing redis clients should be discouraged. Since even adding a Ruby mutex to the blocking code in redis-semaphore would not cover all possible "abuses"*, for now I won't add code to support this use-case.

I'll add a README part that explains it and warns against it. Also, it's still possible as long as you yourself take care 

*: It would only cover the use case where the redis client is only used by redis-semaphore. As soon as other code also uses it the Ruby mutex will be useless.

**Comment by jeremyhaile** [comment]
Is this issue (and the README) out-of-date? The redis-rb gem is threadsafe now and the original "failing" test case works fine for me when testing with the same redis connection.

**Comment by jeremyhaile** [comment]
Never mind...I got it to lock up now...

**Comment by atomical** [comment]
Is the solution is to open a new redis connection in each thread?

**Comment by jcavalieri** [comment]
I'm seeing deadlocks even with new connections. Is there any reason why this library couldn't use a `Mutex` per semaphore key? I'm experimenting with an LRU cache of mutexes.
